packages/headless-react/src/ssr-commerce/providers.tsx

```diff
@@ -0,0 +1,103 @@
+'use client';
+
+import {
+  InferHydratedState,
+  InferStaticState,
+  NavigatorContext,
+} from '@coveo/headless/ssr-commerce';
+import {PropsWithChildren, useEffect, useState} from 'react';
+import {defineCommerceEngine} from './commerce-engine.js';
+
+interface WithDefinitionProps {
+  staticState: InferStaticState<RealDefinition>;
+  navigatorContext: NavigatorContext;
+}
+
+type LooseDefinition = {
+  setNavigatorContextProvider: unknown;
+  build: unknown;
+  hydrateStaticState: unknown;
+  fetchStaticState: unknown;
+  HydratedStateProvider: unknown;
+  StaticStateProvider: unknown;
+};
+
+type RealDefinition =
+  | ReturnType<typeof defineCommerceEngine>['recommendationEngineDefinition']
+  | ReturnType<typeof defineCommerceEngine>['listingEngineDefinition']
+  | ReturnType<typeof defineCommerceEngine>['searchEngineDefinition']
+  | ReturnType<typeof defineCommerceEngine>['standaloneEngineDefinition'];
+
+export function buildProviderWithDefinition(looseDefinition: LooseDefinition) {
+  return function WrappedProvider({
+    staticState,
+    navigatorContext,
+    children,
+  }: PropsWithChildren<WithDefinitionProps>) {
+    const definition = looseDefinition as RealDefinition;
+    type RecommendationHydratedState = InferHydratedState<typeof definition>;
+    const [hydratedState, setHydratedState] = useState<
+      RecommendationHydratedState | undefined
+    >(undefined);
+
+    definition.setNavigatorContextProvider(() => navigatorContext);
+
+    useEffect(() => {
```

Comment:
pas le choix de tout ces any?

Reply:
There is no way to remove all of the any mentions in the useEffect because `.controllers` is an inferred type. There would be casting either way even if we took the controllers as generics, created types to handle `ReturnType<typeof defineCommerceEngine>['standaloneEngineDefinition'];` and pass the controllers generic to it. There still would be plenty of casting everywhere. I think this method is very simple and simple to read. The generics were getting in my way.

The important typing part here is that implementers get errors when using buildProviderWithDefinition(listing:...). This is why I made the `LooseDefinition` type. I think it is enough structure for the user to not make mistakes. This function does not need any fancy inference. It only takes in an input and returns an output. For the real typing inside the function we use  `ReturnType<typeof defineCommerceEngine>['searchEngineDefinition']` so that we use the same types internally and so it errors out when making improper changes to defineCommerceEngine.